File: version1.cpp
#include "version1.h"
#include <sstream>
#include <iostream>

void version1()
{
  string input;
  int *set1;
  int *set2;
  int m = 0;
  int n = 0;
  bool target = true;
  bool flag = true;

  while(flag)
    {
      cout << "1. Enter sets\n";
      cout << "2. Union\n";
      cout << "3. Difference\n";
      cout << "4. Intersection\n";
      cout << "5. Return to main menu\n";
      cout << "Please choose a selection: ";
      getline(cin, input);
      switch(input[0])
	{
	case '1':
	  if(target == false)
	    {
	      if(m > 0)
		{
		  delete set1;
		  set1 = NULL;
		}
	      if(n > 0)
		{
		  delete set2;
		  set2 = NULL;
		}
	      m = 0;
	      n = 0;

	    }
	  createSet(set1, set2, m, n);
	  target = false;
	  break;
	case '2':
	  if(target)
	    {
	      cout << "Please enter set information first...\n";
	      break;
	    }
	  Union(set1, set2, m, n);
	  break;
	case '3':
	  if(target)
	    {
	      cout << "Please enter set information first...\n";
	      break;
	    }
	  Difference(set1, set2, m, n);
	  break;
	case '4':
	  if(target)
	    {
	      cout << "Please enter set information first...\n";
	      break;
	    }
	  Intersection(set1, set2, m, n);
	  break;
	case '5':
	  cout << "Exiting version 1...\n";
	  flag = false;
	  if(m > 0)
	    {
	      delete set1;
	      set1 = NULL;
	    }
	  if(n > 0)
	    {
	      delete set2;
	      set2 = NULL;
	    }
	  break;
	default:
	  cout << "Invalid selection. Please try again...\n";
	}
    }
}


void createSet(int* &x, int* &y, int &m, int &n)
{
  bool flag = true;
  int counter = 1;
  string input;
  string temp;
  int num = 0;
  string input2;
  //counting number of integers are given for both set 1 and set 2 and assigning
  // them to m and n respectively
  while(flag)
    {
      stringstream ss;
      
      cout << "Please enter integers for set " << counter << " separated by spaces: ";
      if(counter == 1)
	{
	  getline(cin, input);
	  ss << input;
	}
      if(counter == 2)
	{
	  getline(cin, input2);
	  ss << input2;
	}
      while(!ss.eof())
	{
	  ss >> temp;
	  if(stringstream(temp) >> num)
	    {
	      if(counter == 1)
		m++;
	      if(counter == 2)
		n++;
	    }
	}
      ++counter;
      if(counter == 3)
	flag = false;
    }
  //cout << "m has a value of : " << m << " input has a value of " << input << endl;
  //cout << "n has a value of : " << n << " input2 has a value of " << input2 << endl;
  
  //populating set 1
  if(m > 0)
    {
      stringstream ss;
      int i = 0;
      int* store = new int[m];
      int k = m;
      bool flag = false;
     
      ss << input;
      while(!ss.eof())
	{
	  flag = false;
	  ss >> temp;
	  if(stringstream(temp) >> num)
	    {
	      for(int j = 0; j < k; j++)
		{
		  if(store[j] == num)
		    {
		      flag = true;
		      --m;
		    }
		}
	    }
	  if(flag == false)
	    {
	      store[i] = num;
	      ++i;
	    }
	}
      x = new int[m];
      for(int i = 0; i < m; i++)
	{
	  x[i] = store[i];
	}
    }
  
  //populating set 2
  if(n > 0)
    {
      stringstream ss;
      int i = 0;
      int* storage = new int[n];
      int k = n;
      bool flag = false;
 
      ss << input2;
      while(!ss.eof())
	{
	  flag = false;
	  ss >> temp;
	  if(stringstream(temp) >> num)
	    {
	      for(int j = 0; j < k; j++)
		{
		  if(storage[j] == num)
		    {
		      flag = true;
		      --n;
		    }
		}
	    }
	  if(flag == false)
	    {
	      storage[i] = num;
	      ++i;
	    }
	}
      y = new int[n];
      for(int i = 0; i < n; i++)
	{
	  y[i] = storage[i];
	}      
    }
  
  //display sets
  cout << "Set 1 contains: {";
  for(int i = 0; i < m; i++)
    {
      cout << x[i];
      if(i + 1 == m);
      else
	cout << ", ";
    }
  cout << "}" << endl;
  cout << "Set 2 contains: {";
  for(int i = 0; i < n; i++)
    {
      cout << y[i];
      if(i + 1 == n);
      else
	cout << ", ";
    }
  cout << "}"<< endl;
  
}
  

void Union(int* x, int* y, int m, int n)
{
  if(m+n == 0)
    {
      cout << "The Union of set 1 and set 2 is { }\n";
      return;
    }
  else if(m == 0)
    {
      cout << "The Union of set 1 and set 2 is {";
      for(int i = 0; i < n; i++)
	{
	  cout << y[i] << ", ";
	}
      cout << "}\n";
      return;
    }
  else if(n == 0)
    {
      cout << "The Union of set 1 and set 2 is {";
      for(int i = 0; i < m; i++)
	{
	  cout << x[i] << ", ";
	}
      cout << "}\n";
    }
  else
    {
      //cout << "m is equal to " << m << endl;
      //cout << "n is equal to " << n << endl;
      int index = 0;
      bool flag = false;
      int* UN = new int[m+n];
      int counter = m+n;
      for(; index < m; index++)
	{
	  UN[index] = x[index];
	}

      for(int i = 0; i < n; i++)
	{
	  for(int j = 0; j < (m+n); j++)
	    {
	      if(UN[j] == y[i])
		{
		  flag = true;
		  counter--;
		}
	    }
	  if(!flag)
	    {
	      UN[index] = y[i];
	      ++index;
	    }
	  flag = false;
	}

      cout << "The Union of set 1 and set 2 is: {";
      for(int i = 0; i < counter; i++)
	{
	  cout << UN[i];
	  if(i+1 == counter)
	    cout << "}";
	  else
	    cout << ", ";
	}
      cout << endl;
    }
}

void Difference(int* x, int* y, int m, int n)
{
  if(m == 0)
    {
      cout << "The Difference between set 1 and set 2 is: {";
      for(int i = 0; i < n; i++)
	{
	  cout << y[i];
	  if(i+1 == n)
	    cout << "}";
	  else
	    cout << ", ";
	}
      cout << endl;
    }
  else if(n == 0)
    {
      cout << "The Difference between set 1 and set 2 is: {";
      for(int i = 0; i < m; i++)
	{
	  cout << x[i];
	  if(i+1 == m)
	    cout << "}";
	  else
	    cout << ", ";
	}
      cout << endl;
    }
  else if(m == 0 && n == 0)
    cout << "The Difference between set 1 and set 2 is : { }";
  else
    {
      int size = m;
      int* DF;
      bool flag = false;
      
      for(int i = 0; i < m; i++)
	{
	  for(int j = 0; j < n; j++)
	    {
	      if(x[i] == y[j])
		--size;
	    }
	}
      //cout << "The difference counter is " << size << endl;
      if(size > 0)
	{
	  DF = new int[size];
	  int index = 0;
	
	  for(int j = 0; j < m; j++)
	    {
	      for(int k = 0; k < n; k++)
		{
		  if(x[j] == y[k])
		    flag = true;
		}
	      if(flag == false)
		{
		  DF[index] = x[j];
		  ++index;
		}
	      flag = false;
	    }
	
	  cout << "The set difference of set 1 and set 2 is: {"; 
	  for(int i = 0; i < size; i++)
	    {
	      cout << DF[i];
	      if(i+1 == size)
		cout << "}";
	      else
		cout << ", ";
	    }
	  cout << endl;
	}
        else
	  {
	    cout << "The set difference of set 1 and set 2 is: { }" << endl;
	  }
    }//end if-else block
}

void Intersection(int* x, int* y, int m, int n)
{
  int* temp;
  int counter = 0;
  for(int i = 0; i < m; i++)
    {
      for(int j = 0; j < n; j++)
	{
	  if(x[i] == y[j])
	    ++counter;
	}
    }
  //cout << "The counter from intersection is : " << counter << endl;
  if(counter == 0)
    {
      cout << "The intersection of set 1 and set 2 is : { }\n";
    }
  else
    {
      bool flag = false;
      int index = 0;
      int i = 0;
      temp = new int[counter];
      
      for(int j = 0; j < m; j++)
	{
	  for(int k = 0; k < n; k++)
	    {
	      if(x[j] == y[k])
		{
		  flag = true;
		  index = j;
		}
	    }
	  if(flag)
	    {
	      temp[i] = x[index];
	      flag = false;
	      ++i;
	    }
	}
      cout << "The intersection of set 1 and set 2 is {";
      for(int i = 0; i < counter; i++)
	{
	  cout << temp[i];
	  if((i+1) == counter)
	    cout << "}\n";
	  else
	    cout << ", ";
	}	 
    }  
}

File: Version1.java
import java.util.List;
import java.util.ArrayList;
import java.util.Scanner;

public class Version1 
{
	public void version1(Scanner sc)
	{
		int input = 0;
		List<Integer> set1 = new ArrayList<Integer>();
		List<Integer> set2 = new ArrayList<Integer>();
		boolean target = true;
		boolean flag = true;

		while(flag)
		{
			System.out.print("1. Enter sets\n");
			System.out.print("2. Union\n");
			System.out.print("3. Difference\n");
			System.out.print("4. Intersection\n");
			System.out.print("5. Return to main menu\n");
			System.out.print("Please choose a selection: ");
			input = sc.nextInt();

			switch(input)
			{
			case 1:
				if(target == false)
				{
					if(set1.size() > 0)
						set1.clear();
					if(set2.size() > 0)
						set2.clear();
				}
				createSet(set1, set2);
				//System.out.println("this is after creating set: ");
				//System.out.println("set1 size = " + set1.size());
				//System.out.println("Set 2 size = " + set2.size());
				target = false;
				break;
			case 2:
				if(target)
				{
					System.out.print("Please enter set information first...\n");
					break;
				}
				Union(set1, set2);
				break;
			case 3:
				if(target)
				{
					System.out.print("Please enter set information first...\n");
					break;
				}
				Difference(set1, set2);
				break;
			case 4:
				if(target)
				{
					System.out.print("Please enter set information first...\n");
					break;
				}
				Intersection(set1, set2);
				break;
			case 5:
				System.out.print("Exiting version 1...\n");
				flag = false;
				if(set1.size() > 0)
					set1.clear();
				if(set2.size() > 0)
					set2.clear();
				return;
			default:
				System.out.print("Invalid selection. Please try again...\n");
			}//end switch
		}//end while
	}//end version 1

	private void Intersection(List<Integer> set1, List<Integer> set2) 
	{
		List<Integer> temp = new ArrayList<Integer>();
		boolean flag = true;

		if(set1.size() > 0 && set2.size() > 0)
		{
			for(int i = 0; i < set1.size(); i++)
			{
				if(set2.contains(set1.get(i)))
					flag = true;
			}
			for(int i = 0; i < set2.size(); i++)
			{
				if(set1.contains(set2.get(i)))
					flag = true;
			}
			if(flag)
			{
				for(int j = 0; j < set1.size(); j++)
				{
					for(int k = 0; k < set2.size(); k++)
					{
						if(set1.get(j) == set2.get(k))
						{
							if(temp.isEmpty())
								temp.add(set1.get(j));
							else
							{
								if(!temp.contains(set1.get(j)))
									temp.add(set1.get(j));
							}
						}
					}
				}
				if(temp.size() == 0)
					System.out.print("The intersection of set 1 and set 2 is { }\n");
				else
				{
					System.out.print("The intersection of set 1 and set 2 is {");
					for(int x = 0; x < temp.size(); x++)
					{
						System.out.print(temp.get(x));
						if((x+1) == temp.size())
							System.out.print("}\n");
						else
							System.out.print(", ");
					}
				}
			} //end if

			else
			{
				System.out.print("The intersection of set 1 and set 2 is : { }\n");
			}
		}
		else 
		{
			System.out.print("The intersection of set 1 and set 2 is : { }\n");
		}
	}

	private void Difference(List<Integer> set1, List<Integer> set2) 
	{
		if(set1.size() == 0)
		{
			System.out.print("The Difference between set 1 and set 2 is: {");
			for(int i = 0; i < set2.size(); i++)
			{
				System.out.print(set2.get(i));
				if(i+1 == set1.size())
					System.out.print("}");
				else
					System.out.print(", ");
			}
			System.out.println();
		}
		else if(set2.size() == 0)
		{
			System.out.print("The Difference between set 1 and set 2 is: {");
			for(int i = 0; i < set1.size(); i++)
			{
				System.out.print(set1.get(i));
				if(i+1 == set1.size())
					System.out.print("}");
				else
					System.out.print(", ");
			}
			System.out.println();
		}
		else if(set1.size() == 0 && set2.size() == 0)
			System.out.print("The Difference between set 1 and set 2 is : { }");
		else
		{
			int size = set1.size();
			List<Integer> DF = new ArrayList<Integer>();
			boolean flag = false;

			if(size > 0)
			{
				for(int j = 0; j < set1.size(); j++)
				{
					for(int k = 0; k < set2.size(); k++)
					{
						if(set1.get(j) == set2.get(k))
							flag = true;
					}
					if(flag == false)
						DF.add(set1.get(j));

					flag = false;
				}

				System.out.print("The set difference of set 1 and set 2 is: {"); 
				for(int i = 0; i < DF.size(); i++)
				{
					System.out.print(DF.get(i));
					if(i+1 == DF.size())
						System.out.print("}");
					else
						System.out.print(", ");
				}
				System.out.println();
			}
			else
			{
				System.out.print("The set difference of set 1 and set 2 is: { }\n");
			}
		}//end if-else block

	}

	private void Union(List<Integer> set1, List<Integer> set2)
	{
		if(set1.size()+set2.size() == 0)
		{
			System.out.print("The Union of set 1 and set 2 is { }\n");
			return;
		}
		else if(set1.size() == 0)
		{
			System.out.print("The Union of set 1 and set 2 is {");
			for(int i = 0; i < set2.size(); i++)
			{
				System.out.print(set2.get(i) + ", ");
			}
			System.out.print("}\n");
			return;
		}
		else if(set2.size() == 0)
		{
			System.out.print("The Union of set 1 and set 2 is {");
			for(int i = 0; i < set1.size(); i++)
			{
				System.out.print(set1.get(i) + ", ");
			}
			System.out.print("}\n");
		}
		else
		{
			int index = 0;
			List<Integer> UN = new ArrayList<Integer>();
			for(; index < set1.size(); index++)
				UN.add(set1.get(index));

			for( int i = 0; i < set2.size(); i++)
			{
				if(UN.contains(set2.get(i)));
				else
					UN.add(set2.get(i));
			}

			System.out.print("The Union of set 1 and set 2 is: {");
			for(int i = 0; i < UN.size(); i++)
			{
				System.out.print(UN.get(i));
				if(i+1 == UN.size())
					System.out.print("}");
				else
					System.out.print(", ");
			}
			System.out.println();
		}

	}

	private void createSet(List<Integer> set1, List<Integer> set2)
	{
		Scanner sc = new Scanner(System.in);
		boolean flag = true;
		int counter = 1;
		List<Integer> s1 = new ArrayList<Integer>();
		List<Integer> s2 = new ArrayList<Integer>();

		//counting number of integers are given for both set 1 and set 2 and assigning
		// them to m and n respectively
		while(flag)
		{
			System.out.print("Please enter integers for set " + counter + " separated by spaces: ");
			if(counter == 1)
			{
				String [] tokens = sc.nextLine().split("\\s");
				for(int i = 0; i < tokens.length; i++)
					s1.add(Integer.parseInt(tokens[i]));

			}
			if(counter == 2)
			{
				String [] tokens = sc.nextLine().split("\\s");
				for(int i = 0; i <tokens.length; i++)
					s2.add(Integer.parseInt(tokens[i]));
			}
			++counter;
			if(counter == 3)
				flag = false;
		}//end while
		
		//populating set 1
		if(s1.size() > 0)
		{
			for(int j = 0; j < s1.size(); j++)
			{
				if(set1.isEmpty())
				{
					set1.add(s1.get(j));
				}
				else
				{
					if(!set1.contains(s1.get(j)))
						set1.add(s1.get(j));
				}

			}
		}
		//populating set 2
		if(s2.size() > 0)
		{
			for(int j = 0; j < s2.size(); j++)
			{
				if(set2.isEmpty())
				{
					set2.add(s2.get(j));
				}
				else
				{
					if(!set2.contains(s2.get(j)))
						set2.add(s2.get(j));
				}

			}
		}
		//display sets
		System.out.print("Set 1 contains: {");
		for(int i = 0; i < set1.size(); i++)
		{
			System.out.print(set1.get(i));
			if(i + 1 == set1.size());
			else
				System.out.print(", ");
		}
		System.out.print("}\n");
		System.out.print("Set 2 contains: {");
		for(int i = 0; i < set2.size(); i++)
		{
			System.out.print(set2.get(i));
			if(i + 1 == set2.size());
			else
				System.out.print(", ");
		}
		System.out.print("}\n");
	}

}//end class version1
